Add handling for post-install and pre-uninstall scripts.

freakboy3742 opened the pull request:
Adds extension points to allow for the injection of a post-install or pre-uninstall script, and for options in the installer.

The value of the pre/post install script doesn't matter for the template; it's treated as a boolean.

The installer options is a list of dictionaries, with 2 keys of significance:
- name: the identifier that is used to 
- description: A short description of the option
- default: true/false

If the configuration defines an option with a name of `foo`, the `OPTION_FOO` property can be used to programmatically drive the installer.

## PR Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
- [x] All new features have been tested
- [x] All new features have been documented
- [x] I have read the **CONTRIBUTING.md** file
- [x] I will abide by the code of conduct

mhsmith commented on {{ cookiecutter.format }}/briefcase.toml:
```diff
@@ -5,6 +5,8 @@ target_version = "0.3.24"
 [paths]
 app_path = "src/app"
 app_packages_path = "src/app_packages"
+install_scripts_path = "src/scripts"
+license_path = "src/LICENSE.rtf"
```
If the app uses `external_package_path`, as Constructor will, then there will be no `src` directory. And `scripts` isn't a good choice for the directory name, because that's one of the directories in the Conda base environment (the equivalent of `bin` on Unix).

One possible solution is to install the scripts to a directory with a name like `_installer` in the root of the installation tree, regardless of whether the app uses `external_package_path` or not. This would also be an appropriate place for the "bootstrap" script mentioned in https://github.com/beeware/briefcase/pull/2496#issuecomment-3369154367.

For `license_path`, I guess the file doesn't need to be in the installation tree, so maybe we can set it to just `LICENSE.rtf`?

mhsmith replied:
> install the scripts to a directory with a name like _installer in the root of the installation tree, regardless of whether the app uses external_package_path or not

As discussed, this means the installation tree would be produced by merging two locations. We could do this manually by copying them into a temporary directory, but maybe WiX can do this by adding additional `Files` or `File` elements.

freakboy3742 replied:
I've pushed an update that makes these changes. LICENSE.rtf is no longer in the source tree; `src` and `extras` are both collated for content to include in the MSI (with `src` being changed to the external package path for an external app).

freakboy3742 replied:
I've made both these changes. LICENSE is no longer included from inside the src directory; and the installer content is merged in from a second file root (`extras`). This way, the final installed content is the package path  (src, or the user-provided external path), plus the content in `extras` for the installer.

freakboy3742 commented on {{ cookiecutter.format }}/{{ cookiecutter.app_name }}.wxs:
```diff
@@ -325,35 +378,58 @@
+        <CustomAction
+            Id="PostInstallAction"
+            Directory="INSTALLFOLDER"
+            ExeCommand="[SystemFolder]cmd.exe /c &quot;[INSTALLFOLDER]scripts\post_install.bat&quot; --allusers=[ALLUSERS]{% if cookiecutter.install_options %}{% for name in cookiecutter.install_options.keys() %} --{{ name }}=[OPTION_{{ name.upper() }}]{% endfor %}{% endif %}"
```
It is if we want to maintain compatibility with the main Briefcase branch. However, since these if statements won't be necessary once this PR lands, I guess we should leave them out and test this branch against the PR.

freakboy3742 commented:
@mhsmith I've introduced a launcher script that captures the command line arguments and converts them into environment variables. 

I've also fixed the ALLUSER handling so it return 0 or 1.

mhsmith commented on {{ cookiecutter.format }}/extras/{{ cookiecutter.installer_path }}/_run_post_install.bat:
```diff
@@ -0,0 +1,14 @@
+SET {{ name.upper() }}=%{{ argn }}
+{% endfor -%}
+
+CALL "%~dp0post_install.bat"
```
Why do the "run" scripts have leading underscores, but the actual post_install script doesn't? We're not using the user's original filename, so there's no chance of a clash.

freakboy3742 replied:
That's fair - my original thinking was picking a name to avoid collisions, but as you note, we're renaming the script, so that's a non-issue.

mhsmith commented on {{ cookiecutter.format }}/{{ cookiecutter.app_name }}.wxs:
```diff
@@ -278,7 +331,7 @@
                 Property="MSIINSTALLPERUSER"
                 Value="1" />
-            {% if cookiecutter.console_app %}
+                {% if cookiecutter.console_app %}
```
Moved to #76 to avoid blocking this PR.
